test/e2e/service.go

```diff
@@ -437,54 +539,95 @@ func validateUniqueOrFail(s []string) {
+	By(fmt.Sprintf("successfully translated pod names to ips: %v -> %v on namespace %s", expectedEndpoints, portsByIp, ns))
+	return portsByIp
 }
 
-func validateEndpointsOrFail(c *client.Client, ns, serviceName string, expectedPort int, expectedPods []string) {
+func validatePortsOrFail(endpoints map[string][]int, expectedEndpoints map[string][]int) {
+	if len(endpoints) != len(expectedEndpoints) {
+		Failf("invalid number of endpoints (should not happen because we check this condition before) got %v, expected %v", endpoints, expectedEndpoints)
+	}
+	for ip := range expectedEndpoints {
+		if _, ok := endpoints[ip]; !ok {
+			Failf("endpoint %v not found", ip)
+		}
+		if len(endpoints[ip]) != len(expectedEndpoints[ip]) {
+			Failf("invalid list of ports for ip %v. Got %v, expected %v", ip, endpoints[ip], expectedEndpoints[ip])
+		}
+		sort.Ints(endpoints[ip])
+		sort.Ints(expectedEndpoints[ip])
+		for index := range endpoints[ip] {
+			if endpoints[ip][index] != expectedEndpoints[ip][index] {
+				Failf("invalid list of ports for ip %v. Got %v, expected %v", ip, endpoints[ip], expectedEndpoints[ip])
+			}
+		}
+	}
+}
+
+func validateEndpointsOrFail(c *client.Client, ns, serviceName string, expectedEndpoints map[string][]int) {
+	expectedPorts := make(map[int]bool)
+	for _, ports := range expectedEndpoints {
+		for _, port := range ports {
+			expectedPorts[port] = true
+		}
+	}
+
 	for {
 		endpoints, err := c.Endpoints(ns).Get(serviceName)
 		if err == nil {
-			ips := flattenSubsets(endpoints.Subsets, expectedPort)
-			if len(ips) == len(expectedPods) {
-				validateIPsOrFail(c, ns, expectedPods, ips)
+			By(fmt.Sprintf("Found endpoints %v", endpoints))
+
+			// I don't really understand why we are ignoring ports that are not expected to be present.
+			// Do we expect to find other ports in the service we created?
+			// I found this check in the previous version of the test and I let it stay here.
+			portsByIp := getPortsByIp(endpoints.Subsets, expectedPorts)
+
+			By(fmt.Sprintf("Found ports by ip %v", portsByIp))
+			if len(portsByIp) == len(expectedEndpoints) {
+				expectedPortsByIp := translatePodNameToIpOfFail(c, ns, expectedEndpoints)
+				validatePortsOrFail(portsByIp, expectedPortsByIp)
 				break
 			} else {
-				By(fmt.Sprintf("Unexpected number of endpoints: found %v, expected %v (ignoring for 1 second)", ips, expectedPods))
+				By(fmt.Sprintf("Unexpected number of endpoints: found %v, expected %v (ignoring for 1 second)", portsByIp, expectedEndpoints))
 			}
 		} else {
 			By(fmt.Sprintf("Failed to get endpoints: %v (ignoring for 1 second)", err))
 		}
 		time.Sleep(time.Second)
 	}
-	By(fmt.Sprintf("successfully validated endpoints %v port %d on service %s/%s", expectedPods, expectedPort, ns, serviceName))
+	By(fmt.Sprintf("successfully validated endpoints %v with ports %v on service %s/%s", expectedEndpoints, expectedPorts, ns, serviceName))
```

Comment:
I added the same message at the beginning. I think it is necessary to print it at the end too, because this function is long and does retries.